File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
// https://www.acmicpc.net/problem/2110

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int n, c;
    static int arr[];

      static boolean check(int dist) {
        int cnt = 1;
        int cur = arr[1];
        for (int i = 2; i <= n; i++) {
            if (cur + dist <= arr[i]) {
                cnt += 1;
                cur = arr[i];
            }
        }

        return c <= cnt;
    }


    static int lower_bound(int arr[], int target) {
        int st = 1, en = 1000000000;
        int result = 0;
        while (st <= en) {
            int mid = (st+en)/ 2;

            if (check(mid)) {
                result = mid;
                st = mid + 1;
            } else {
                en = mid-1;
            }
        }

        return result;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        arr = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        Arrays.sort(arr, 1, n + 1);
        int result = lower_bound(arr, c);
        System.out.println(result);
    }
}

File: 백준/Gold/2110. 공유기 설치/공유기 설치.py

def decide(dist):
    cnt = 1
    cur = arr[0]
    for i in range(1,N):
        if cur + dist <= arr[i]:
            cnt += 1
            cur = arr[i]

    return C <= cnt

def lower_bound(arr):
    l = 1
    r = 1_000_000_000;
    ret = 0
    while l<=r :
        mid = (l+r) // 2
        if(decide(mid)):
            ret = mid
            l = mid + 1
        else:
            r = mid - 1

    return ret

if __name__=="__main__":
    N,C = map(int,input().split())
    arr = list()
    for _ in range(N):
        arr.append(int(input()))

    arr.sort()
    print(lower_bound(arr))